common/persistence/nosql/shardingPolicy.go

```diff
@@ -0,0 +1,111 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nosql
+
+import (
+	"fmt"
+
+	"github.com/uber/cadence/common/log/tag"
+
+	"github.com/dgryski/go-farm"
+
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/log"
+)
+
+type (
+	// shardingPolicy holds the configuration for the sharding logic
+	shardingPolicy struct {
+		logger       log.Logger
+		defaultShard string
+		config       config.ShardedNoSQL
+
+		hasShardedHistory  bool
+		hasShardedTasklist bool
+	}
+)
+
+func newShardingPolicy(logger log.Logger, cfg config.ShardedNoSQL) (shardingPolicy, error) {
+	sp := shardingPolicy{
+		logger:       logger,
+		defaultShard: cfg.DefaultShard,
+		config:       cfg,
+	}
+
+	err := sp.parse()
+	if err != nil {
+		return sp, err
+	}
+
+	return sp, nil
+}
+
+func (sp *shardingPolicy) parse() error {
+	sp.parseHistoryShardMapping()
+	sp.parseTaskListShardingPolicy()
+	return nil
+}
+
+func (sp *shardingPolicy) parseHistoryShardMapping() {
+	historyShardMapping := sp.config.ShardingPolicy.HistoryShardMapping
+	if len(historyShardMapping) == 0 {
+		return
+	}
+	sp.hasShardedHistory = true
+}
+
+func (sp *shardingPolicy) parseTaskListShardingPolicy() {
+	tlShards := sp.config.ShardingPolicy.TaskListHashing.ShardOrder
+	if len(tlShards) == 0 {
+		return
+	}
+	sp.hasShardedTasklist = true
+}
+
+func (sp *shardingPolicy) getHistoryShardName(shardID int) string {
+	if !sp.hasShardedHistory {
+		sp.logger.Debug("Selected default store shard for history shard", tag.StoreShard(sp.defaultShard), tag.ShardID(shardID))
+		return sp.defaultShard
+	}
+
+	for _, r := range sp.config.ShardingPolicy.HistoryShardMapping {
+		if shardID >= r.Start && shardID <= r.End {
+			sp.logger.Debug("Selected store shard history shard", tag.StoreShard(r.Shard), tag.ShardID(shardID))
+			return r.Shard
+		}
+	}
+
+	panic(fmt.Sprintf("Failed to identify store shard for shardID %v", shardID))
+}
+
+func (sp *shardingPolicy) getTaskListShardName(domainID string, taskListName string, taskType int) string {
+	if !sp.hasShardedTasklist {
+		sp.logger.Debug("Selected default store shard for tasklist", tag.StoreShard(sp.defaultShard), tag.WorkflowTaskListName(taskListName))
+		return sp.defaultShard
+	}
+	tlShards := sp.config.ShardingPolicy.TaskListHashing.ShardOrder
+	tlShardCount := len(tlShards)
+	hash := farm.Hash32([]byte(domainID+"_"+taskListName)) % uint32(tlShardCount)
```

Comment:
having just made a silly mistake with this myself, just double-cehcking, are there any conditions under which tlShardCount could be zero? https://go.dev/play/p/Pv73y1KpbOZ

Reply:
It should be impossible with the `if !sp.hasShardedTasklist` at the top of the func.